### DISCUSSION TITLE
Red screen Unauthorized SEB

### DISCUSSION THREAD

**Discussion by jladders** [SafeExamBrowser/seb-mac#372] [General]
![Image](https://github.com/SafeExamBrowser/seb-mac/assets/146181235/c2c3bae0-310e-4c00-a014-9459e70890b0)

Hello,

We have packaged this and uploaded via Jamf and tested. Today we are getting this error and we don't know why. 

We have downloaded from the source which is a .dmg and re done as a .pkg for Jamf. 

Any help would be great. 

Thanks

**Comment by danschlet** [comment]
If the code signature of SEB is changed, then this error will be displayed, also the BrowserExamKey and AppSignatureKey (ASK) will be wrong. Make sure packaging does not at all alter the Safe Exam Browser.app bundle (also NOT adding any files to the bundle, hidden or not). You can use [SEB Verificator](https://github.com/SafeExamBrowser/seb-mac-verificator/releases/download/SEBVerificator_1.0.2/SEBVerificator-1.0.2.dmg) to check if a SEB app version has been altered. 

Note that if some tool removes the unused architecture (Intel or ARM) from the SEB Universal Binary, the code signature won't be altered, but a wrong ASK will be displayed in SEB Server.

**Reply by jladders** [reply]
Thank you for your quick reply. Will check

**Reply by jladders** [reply]
as its a .DMG will having it as a .PKG and signing It cause the issue?

**Reply by danschlet** [reply]
I don't know if creating a .pkg from a .dmg changes the code signature of the installed SEB .app bundle. It looks like, but that would be extremely stupid to do for any packaging or deployment tool! The code signature of an app exists for the reason to identify an app, to prove its original developer. So if any tool thinks it's a good idea to resign apps that's insanity and not at all good practice! Maybe contact your IT department (the once who are packaging apps) or the vendor(s) of the tool(s) used and ask them, if packaged apps are resigned. Feel free to tell them to contact us (so I can scold them for this insanity if they're really resigning apps). 

It's of course also possible that there is some macOS or even SEB bug, although I don't expect latter, but maybe it could be some unexpected edge case.

**Reply by danschlet** [reply]
You can check the code signature of an installed app in Terminal with this command: `codesign -dv --verbose=4 <path of .app>`. This should be displayed with an original SEB:

```
codesign -dv --verbose=4 /Applications/Safe\ Exam\ Browser.app 
Executable=/Applications/Safe Exam Browser.app/Contents/MacOS/Safe Exam Browser
Identifier=org.safeexambrowser.SafeExamBrowser
Format=app bundle with Mach-O universal (x86_64 arm64)
CodeDirectory v=20500 size=21711 flags=0x10000(runtime) hashes=667+7 location=embedded
VersionPlatform=1
VersionMin=658688
VersionSDK=852736
Hash type=sha256 size=32
CandidateCDHash sha256=1792ea63aeaebf90b7a0dbbcdf0c305e68a14209
CandidateCDHashFull sha256=1792ea63aeaebf90b7a0dbbcdf0c305e68a142090c9fc67845b215821a3a2bfd
Hash choices=sha256
CMSDigest=1792ea63aeaebf90b7a0dbbcdf0c305e68a142090c9fc67845b215821a3a2bfd
CMSDigestType=2
Executable Segment base=0
Executable Segment limit=2097152
Executable Segment flags=0x1
Page size=4096
Launch Constraints:
	None
CDHash=1792ea63aeaebf90b7a0dbbcdf0c305e68a14209
Signature size=8981
Authority=Developer ID Application: Daniel Schneider (6F38DNSC7X)
Authority=Developer ID Certification Authority
Authority=Apple Root CA
Timestamp=13 Jan 2024 at 00:04:14
Info.plist entries=36
TeamIdentifier=6F38DNSC7X
Runtime Version=13.3.0
Sealed Resources version=2 rules=13 files=95
Internal requirements count=1 size=228
```

Important are these three lines, which should not be different:
```
Authority=Developer ID Application: Daniel Schneider (6F38DNSC7X)
Authority=Developer ID Certification Authority
Authority=Apple Root CA
```

Can you please post what this command shows for the SEB which is installed on the affected machine?

**Reply by jladders** [reply]
Hi Dan, 

I think because it's a .DMG it doesn't work with Jamf well in terms of deploying to over 100 iMacs. As far as I am aware this got repackaged as a .PKG and got re signed for the installer part so we had no gatekeeper issues. Unless you have a version which is a .pkg?

**Reply by danschlet** [reply]
Did this really change the code signature of the installed app bundle? Can you send me the output of the above mentioned Terminal command (you can also use info at safeexambrowser dot org for privacy reasons).

**Reply by danschlet** [reply]
~~Hmm, thinking this through probably it's really impossible to not change the app signature when repackaging it into a .pkg installer~~. Note: It is possible, I tried it out.

That's still quite annoying though and also why a software like Jamf cannot directly distribute app bundles (from .dmg disk images or whatever source). 

I guess then we will really have to also provide .pkg versions of SEB releases. I'm trying to avoid any extra efforts, as our resources are limited and any additional format potentially can create new sources of error or support questions. But I guess we'll have to invest that effort...

**Reply by jladders** [reply]
I am just waiting to get back into the suite and run those diagnostics. Going forward a .pkg would really be helpful as stuff like this is deployed via Jamf to over 100 machines

**Reply by jladders** [reply]
Just sent the email to you now

**Reply by jladders** [reply]
Noticed 3.2.5 was installed. Jamf then sent down 3.3.2 and then the issues occurred. Have done some testing. Jamf pushes an update over the top which is causing the issue. When you uninstall and then installed via Jamf, the SEB Verificator comes up ok

**Reply by danschlet** [reply]
Perfect, thanks for this update!

So the issue in this case seems to have been that the MDM solution you are using didn't remove all resources of the previously installed SEB version (3.2.5) and probably just added the resource files of SEB 3.3.2, overwriting the old resource files with the same name, but keeping the some no longer existing resources of SEB 3.2.5 in the app bundle. This apparently invalidated the code signature of the SEB app.

I don't understand how this is possible with a professional solution like Jamf Pro, maybe it had to do with the fact that you have to create a .pkg file for Jamf MDM deployment and that pkg installer did a kind of incremental update instead of replacing the whole app bundle (which basically is a folder, but macOS interprets it as an app).

How did you in the end forced Jamf to first uninstall the old SEB version? Did you first execute an uninstall action, then an install action (don't remember the right term used in Jamf, but I guess you know what I mean)? Or did you change the .pkg file?

**Comment by danschlet** [comment]
Might be related to: https://github.com/SafeExamBrowser/seb-mac/issues/376
